### PR TITLE
Cleanup and extension of Spec and Unit queries

### PR DESCRIPTION
**Pull request by pawelbaran**
### Issues addressed by this PR

Closes #

### Changelog

- renamed `UnitTypeFromPropertyName` to `SpecFromName` and `UnitTypePropertyName` to `SpecName`
- added `UnitName` and `UnitFromName`

### CODE REVIEW

**Comment by pawelbaran** [comment]
@BHoMBot check compliance

**Comment by pawelbaran** [comment]
@BHoMBot check compliance

File: Revit_Core_Engine/Query/UnitFromName.cs [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+#endif
+        }
+
+
+        /***************************************************/
+        /****              Private methods              ****/
+        /***************************************************/
+
+#if (REVIT2020)
+        private static void CollectUnits()
```
Comment (vietle-bh) [inline Revit_Core_Engine/Query/UnitFromName.cs:70]:
```suggestion
        private static void CollectUnitTypeIds()
```

File: Revit_Core_Engine/Query/UnitFromName.cs [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+        /***************************************************/
+        /****               Private fields              ****/
+        /***************************************************/
+
+        private static Dictionary<string, DisplayUnitType> m_UnitsWithNames = null;
+
+        /***************************************************/
+#else
+
+        private static void CollectUnits()
```
Comment (vietle-bh) [inline Revit_Core_Engine/Query/UnitFromName.cs:94]:
```suggestion
        private static void CollectUnitTypeIds()
```

File: Revit_Core_Engine/Query/UnitFromName.cs [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+        /****              Public methods               ****/
+        /***************************************************/
+
+        [Description("Returns Revit unit object (enum for Revit up to 2020 or ForgeTypeId for later versions) based on UnitTypeId property name that represents it.")]
+        [Input("name", "Name of UnitTypeId property to be queried for the correspondent unit.")]
+        [Output("unit", "Unit object under the input UnitTypeId property name.")]
+#if (REVIT2020)
+        public static DisplayUnitType UnitFromName(this string name)
+#else
+        public static ForgeTypeId UnitFromName(this string name)
```
Comment (vietle-bh) [inline Revit_Core_Engine/Query/UnitFromName.cs:43]:
```suggestion
        public static DisplayUnitType UnitTypeId(this string name)
#else
        public static ForgeTypeId UnitTypeId(this string name)
```

File: Revit_Core_Engine/Query/UnitName.cs [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+        /****              Public methods               ****/
+        /***************************************************/
+
+        [Description("Returns name of UnitTypeId property that contains a given Revit unit object (enum for Revit up to 2020 or ForgeTypeId for later versions).")]
+        [Input("unit", "Unit object to be queried for the correspondent UnitTypeId property name.")]
+        [Output("name", "Name of UnitTypeId property that contains the input unit object.")]
+#if (REVIT2020)
+        public static string UnitName(this DisplayUnitType unit)
+#else
+        public static string UnitName(this ForgeTypeId unit)
```
Comment (vietle-bh) [inline Revit_Core_Engine/Query/UnitName.cs:42]:
```suggestion
        public static string UnitTypeName(this DisplayUnitType unit)
#else
        public static string UnitTypeName(this ForgeTypeId unit)
```

File: Revit_Core_Engine/Query/UnitName.cs [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+            else
+                return null;
+        }
+
+        /***************************************************/
+
+        [Description("Returns name of UnitTypeId property that contains given Revit parameter's unit object (enum for Revit up to 2020 or ForgeTypeId for later versions).")]
+        [Input("parameter", "Parameter to be queried for the correspondent UnitTypeId property name.")]
+        [Output("name", "Name of UnitTypeId property that contains the unit object correspondent to the input parameter.")]
+        public static string UnitName(this Parameter parameter)
```
Comment (vietle-bh) [inline Revit_Core_Engine/Query/UnitName.cs:59]:
```suggestion
        public static string UnitTypeName(this Parameter parameter)
```

File: Revit_Core_Engine/Query/SpecName.cs [hunk 6]
```diff
@@ -33,32 +33,35 @@ public static partial class Query
...
-        [Input("unitType", "Unit type object to be queried for the correspondent SpecTypeId property name.")]
-        [Output("identifier", "Name of SpecTypeId property that contains the input unit type object.")]
+        [PreviousVersion("7.3", "BH.Revit.Engine.Core.Query.UnitTypePropertyName(Autodesk.Revit.DB.UnitType)")]
+        [PreviousVersion("7.3", "BH.Revit.Engine.Core.Query.UnitTypePropertyName(Autodesk.Revit.DB.ForgeTypeId)")]
+        [Description("Returns name of SpecTypeId property that contains a given Revit spec object (enum for Revit up to 2020 or ForgeTypeId for later versions).")]
+        [Input("spec", "Spec object to be queried for the correspondent SpecTypeId property name.")]
+        [Output("name", "Name of SpecTypeId property that contains the input spec object.")]
 #if (REVIT2020)
-        public static string UnitTypePropertyName(this UnitType unitType)
+        public static string SpecName(this UnitType spec)
```
Comment (vietle-bh) [inline Revit_Core_Engine/Query/SpecName.cs:42]:
```suggestion
        public static string SpecTypeName(this UnitType spec)
```

File: Revit_Core_Engine/Query/SpecName.cs [hunk 7]
```diff
@@ -33,32 +33,35 @@ public static partial class Query
...
 #else
-        public static string UnitTypePropertyName(this ForgeTypeId unitType)
+        public static string SpecName(this ForgeTypeId spec)
```
Comment (vietle-bh) [inline Revit_Core_Engine/Query/SpecName.cs:44]:
```suggestion
        public static string SpecTypeName(this ForgeTypeId spec)
```

File: Revit_Core_Engine/Query/SpecName.cs [hunk 8]
```diff
@@ -33,32 +33,35 @@ public static partial class Query
...
         /***************************************************/
 
-        [Description("Returns name of SpecTypeId property that contains given Revit parameter's unit type object (enum for Revit up to 2020 or ForgeTypeId for later versions).")]
+        [PreviousVersion("7.3", "BH.Revit.Engine.Core.Query.UnitTypePropertyName(Autodesk.Revit.DB.Parameter)")]
+        [Description("Returns name of SpecTypeId property that contains given Revit parameter's spec object (enum for Revit up to 2020 or ForgeTypeId for later versions).")]
         [Input("parameter", "Parameter to be queried for the correspondent SpecTypeId property name.")]
-        [Output("identifier", "Name of SpecTypeId property that contains the unit type object correspondent to the input parameter.")]
-        public static string UnitTypePropertyName(this Parameter parameter)
+        [Output("name", "Name of SpecTypeId property that contains the spec object correspondent to the input parameter.")]
+        public static string SpecName(this Parameter parameter)
```
Comment (vietle-bh) [inline Revit_Core_Engine/Query/SpecName.cs:62]:
```suggestion
        public static string SpecTypeName(this Parameter parameter)
```

File: Revit_Core_Engine/Query/SpecFromName.cs [hunk 9]
```diff
@@ -35,24 +34,25 @@ public static partial class Query
...
-        [Description("Returns Revit unit type object (enum for Revit up to 2020 or ForgeTypeId for later versions) based on SpecTypeId property name that represents it.")]
-        [Input("identifier", "Name of SpecTypeId property to be queried for the correspondent unit type.")]
-        [Output("unitType", "Unit type object under the input SpecTypeId property name.")]
+        [PreviousVersion("7.3", "BH.Revit.Engine.Core.Query.UnitTypeByPropertyName(System.String)")]
+        [Description("Returns Revit spec object (enum for Revit up to 2020 or ForgeTypeId for later versions) based on SpecTypeId property name that represents it.")]
+        [Input("name", "Name of SpecTypeId property to be queried for the correspondent spec.")]
+        [Output("spec", "Spec object under the input SpecTypeId property name.")]
 #if (REVIT2020)
-        public static UnitType UnitTypeByPropertyName(this string identifier)
+        public static UnitType SpecFromName(this string name)
```
Comment (vietle-bh) [inline Revit_Core_Engine/Query/SpecFromName.cs:42]:
```suggestion
        public static UnitType SpecTypeId(this string name)
```

File: Revit_Core_Engine/Query/SpecFromName.cs [hunk 10]
```diff
@@ -35,24 +34,25 @@ public static partial class Query
...
 #else
-        public static ForgeTypeId UnitTypeByPropertyName(this string identifier)
+        public static ForgeTypeId SpecFromName(this string name)
```
Comment (vietle-bh) [inline Revit_Core_Engine/Query/SpecFromName.cs:44]:
```suggestion
        public static ForgeTypeId SpecTypeId(this string name)
```

**Review by vietle-bh — COMMENTED** [review]
All my comments are for the same idea of making the return types more explicit. I actually find the Revit API's new unit classes very confusing so this would help me in using these methods in the future. Future devs can benefit too 👍

**Comment by vietle-bh** [comment]
Sorry, I just saw in another context that you wanted to "ditch naming Unit Type and Display Unit Type altogether". While `DisplayUnitType` should go, the latest Revit API still has `UnitTypeId` so I guess we can still use the term `UnitType` where it makes sense.

**Comment by pawelbaran** [comment]
> Sorry, I just saw in another context that you wanted to "ditch naming Unit Type and Display Unit Type altogether". While `DisplayUnitType` should go, the latest Revit API still has `UnitTypeId` so I guess we can still use the term `UnitType` where it makes sense.

The problem is that the old `DisplayUnitType` is correspondent to `UnitTypeId`, while `UnitType` can be taken from `SpecTypeId` class. So, besides renaming stuff, Autodesk also managed to reuse the old concept of 'Unit Type' in the new context 🤯 This is why I got so much into 'Spec' and 'Unit' to avoid ambiguity with the old times.

**Comment by vietle-bh** [comment]
> The problem is that the old `DisplayUnitType` is correspondent to `UnitTypeId`, while `UnitType` can be taken from `SpecTypeId` class. So, besides renaming stuff, Autodesk also managed to reuse the old concept of 'Unit Type' in the new context 

I think we should only consider the new system going forward, as `DisplayUnitType` can be removed from our code within a year, while being descriptive on the methods can make them more useful for a long time 👍

**Comment by pawelbaran** [comment]
> All my comments are for the same idea of making the return types more explicit. I actually find the Revit API's new unit classes very confusing so this would help me in using these methods in the future. Future devs can benefit too 👍
> 
> 

Thanks for so much thought put into the comments - indeed, the new system is so confusing. I can see that you suggest adding `Type` suffix to specs and units. Not sure, however, if these are needed actually- according to Revit API documentation [`SpecTypeId` class _contains constants identifying specs_](https://www.revitapidocs.com/2023/35c5a3db-b4b3-daa0-ebdd-ae87d03cd5bb.htm), i.e. the base concept is `Spec`, `TypeId` being noise added to name the class in a complex way 🙈 So what I am trying to do is demistyfying the whole thing and simplifying it to _Spec_ and _Unit_.

Another thing that I tried to achieve with chosen naming was coupling _to_ and _from_: `SpecName` query to get the name from a spec, coupled with `SpecFromName` to get the spec from its name. With your suggestion, we lose this coupling, which I think is worth keeping.

What do you think?

**Comment by vietle-bh** [comment]
Yes, let's keep your naming system. It's definitely better 👍

**Review by vietle-bh — APPROVED** [review]
Tested and it worked 👍

**Comment by pawelbaran** [comment]
@BHoMBot check core
@BHoMBot check serialisation
@BHoMBot check null-handling

**Comment by pawelbaran** [comment]
@BHoMBot check installer

**Comment by pawelbaran** [comment]
@BHoMBot check ready-to-merge

**Pull request merged**